File: testdata/fixtures/csharp/CWE-327.cs
// CWE-327/328: Weak hash — MD5/SHA1 have practical collisions.
// MD5, SHA1, or HashAlgorithm.Create("MD5") are flagged; they are
// unsafe for signatures or fingerprints. Fix with SHA256 or stronger.
using System.Security.Cryptography;

namespace Fixture {
    public class Hashing {
        public byte[] Fingerprint(byte[] data) {
            // ruleid: csharp-weak-hash-algorithm
            using (var h = MD5.Create()) {
                return h.ComputeHash(data);
            }
        }

        public byte[] Legacy(byte[] data) {
            // ruleid: csharp-weak-hash-algorithm
            using (var h = SHA1.Create()) {
                return h.ComputeHash(data);
            }
        }

        public byte[] Named(byte[] data) {
            // ruleid: csharp-weak-hash-algorithm
            using (var h = HashAlgorithm.Create("MD5")) {
                return h.ComputeHash(data);
            }
        }

        public byte[] Explicit(byte[] data) {
            // ruleid: csharp-weak-hash-algorithm
            using (var h = new MD5CryptoServiceProvider()) {
                return h.ComputeHash(data);
            }
        }

        public byte[] Strong(byte[] data) {
            // ok: csharp-weak-hash-algorithm
            using (var h = SHA256.Create()) {
                return h.ComputeHash(data);
            }
        }
    }
}

File: testdata/fixtures/elixir/CWE-327.ex
# CWE-327: weak crypto
defmodule Fixture.CWE327 do
  def vuln(data) do
    # ruleid: elixir-weak-crypto
    :md5.hash(data)
    # ruleid: elixir-weak-crypto
    :crypto.hash(:md5, data)
    # ruleid: elixir-weak-crypto
    :crypto.hash(:sha1, data)
  end
  def safe(data) do
    # ok: elixir-weak-crypto
    :crypto.hash(:sha256, data)
    # ok: elixir-weak-crypto
    :crypto.hash(:sha3_256, data)
  end
end
